### PR TITLE
[Enhancement]Avoid unnecessary calculations in HashTableGrower

### LINKED ISSUE #9224: [Enhancement] Avoid unnecessary calculations in `HashTableGrower` & `HashTableFixedGrower`
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Doris use bit operation In `HashTableGrower` & `HashTableFixedGrower` to get buffer size like this:
```cpp
size_t buf_size() const { return 1ULL << size_degree; }
```
The operation will be executed every time when `buf_size()` is called, but this is not necessary.

### Solution

To remove unnecessary calculations, we can save the result in a newly added field.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by mrhhsg**
# Proposed changes

Issue Number: close #9224 

## Problem Summary:

After using this enhancement, the time to build the hash table is slightly reduced.
I test this with TPC-H Q1 with 10GB and run Q1 4 times separately.
|seq|Without the enhancement|With the enhancement|
|--|---------|----------|
|1|3s480ms|3s449ms|
|2|3s274ms|3s277ms|
|3|3s283ms|3s273ms|
|4|3s331ms|3s281ms|

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/common/hash_table/hash_table.h [hunk 1]
```diff
@@ -245,11 +245,14 @@ struct HashTableGrower {
     /// The state of this structure is enough to get the buffer size of the hash table.
     doris::vectorized::UInt8 size_degree = initial_size_degree;
 
+    size_t _buf_size = 1ULL << initial_size_degree;
+    size_t _max_fill = 1ULL << (initial_size_degree - 1);
+
     /// The size of the hash table in the cells.
-    size_t buf_size() const { return 1ULL << size_degree; }
+    size_t buf_size() const { return _buf_size; }
```
Comment (yiguolei) [inline be/src/vec/common/hash_table/hash_table.h:252]: if the size_degree is modified in increase_size method. The buffer size should changed but in your code, the buffer size is const.

File: be/src/vec/common/hash_table/hash_table.h [hunk 2]
```diff
@@ -245,11 +245,14 @@ struct HashTableGrower {
     /// The state of this structure is enough to get the buffer size of the hash table.
     doris::vectorized::UInt8 size_degree = initial_size_degree;
 
+    size_t _buf_size = 1ULL << initial_size_degree;
+    size_t _max_fill = 1ULL << (initial_size_degree - 1);
```
Comment (yiguolei) [inline be/src/vec/common/hash_table/hash_table.h:249]: I think we'd better not change this. I find that it only improved about 0.02%. But the readability of the code is affected. Maybe @HappenLee  could read this PR.

**Comment by yangzhg** [comment]
The query performance change looks very small, more like normal query time consuming jitter, are you sure this pr has a performance improvement?

**Comment by mrhhsg** [comment]
> The query performance change looks very small, more like normal query time consuming jitter, are you sure this pr has a performance improvement?

hi @yangzhg 
In fact, the impact is very small, but a bit operation will be performed each time when data is inserted into the hash table.
For example, it may take tens of milliseconds to perform 150,000 bit operations.

**Comment by yiguolei** [comment]
I will close this PR since it only have 0.02% improvement.

**Pull request closed without merging**